sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/DefaultSdkMeterProvider.java

```diff
@@ -0,0 +1,122 @@
+import io.opentelemetry.sdk.internal.ComponentRegistry;
+import io.opentelemetry.sdk.metrics.data.MetricData;
+import io.opentelemetry.sdk.metrics.exemplar.ExemplarSampler;
+import io.opentelemetry.sdk.metrics.export.MetricProducer;
+import io.opentelemetry.sdk.metrics.export.MetricReader;
+import io.opentelemetry.sdk.metrics.internal.export.CollectionHandle;
+import io.opentelemetry.sdk.metrics.internal.state.MeterProviderSharedState;
+import io.opentelemetry.sdk.metrics.internal.view.ViewRegistry;
+import io.opentelemetry.sdk.resources.Resource;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * {@code SdkMeterProvider} implementation for {@link MeterProvider}.
+ *
+ * <p>This class is not intended to be used in application code and it is used only by {@link
+ * OpenTelemetry}.
+ *
+ * <p>WARNING: A MetricProducer is stateful. It will only return changes since the last time it was
+ * accessed. This means that if more than one {@link
+ * io.opentelemetry.sdk.metrics.export.MetricExporter} has a handle to this MetricProducer, the two
+ * exporters will not receive copies of the same metric data to export.
+ */
+final class DefaultSdkMeterProvider implements SdkMeterProvider {
+
+  private static final Logger LOGGER = Logger.getLogger(DefaultSdkMeterProvider.class.getName());
+  static final String DEFAULT_METER_NAME = "unknown";
+  private final ComponentRegistry<SdkMeter> registry;
+  private final MeterProviderSharedState sharedState;
+
+  // These are *only* mutated in our constructor, and safe to use concurrently after construction.
+  private final Set<CollectionHandle> collectors = CollectionHandle.mutableSet();
+  private final List<MetricReader> readers = new ArrayList<>();
+
+  DefaultSdkMeterProvider(
+      List<MetricReader.Factory> readerFactories,
+      Clock clock,
+      Resource resource,
+      ViewRegistry viewRegistry,
+      ExemplarSampler exemplarSampler) {
+    this.sharedState =
+        MeterProviderSharedState.create(clock, resource, viewRegistry, exemplarSampler);
+    this.registry =
+        new ComponentRegistry<>(
+            instrumentationLibraryInfo -> new SdkMeter(sharedState, instrumentationLibraryInfo));
+
+    // Here we construct our own unique handle ids for this SDK.
+    // These are guaranteed to be unique per-reader for this SDK, and only this SDK.
+    Supplier<CollectionHandle> handleSupplier = CollectionHandle.createSupplier();
+    for (MetricReader.Factory readerFactory : readerFactories) {
+      CollectionHandle handle = handleSupplier.get();
+      // TODO: handle failure in creation or just crash?
+      MetricReader reader = readerFactory.apply(new LeasedMetricProducer(handle));
+      collectors.add(handle);
+      readers.add(reader);
+    }
```

Comment:
I don't much like doing all this work in the constructor. Can we move this stuff to an `init` method of some sort to be called after construction is complete?

Reply:
Would this be called immediately after construction always anyway?

If we want immutable / final references to this config, it needs to be in the constructor.  I moved it this way to avoid having a "register" method be publicly exposed the could dynamically adjust configuration (and require locks).   We can shift this around, but having another method that mutates state feels just as bad as exposing "register" to me.